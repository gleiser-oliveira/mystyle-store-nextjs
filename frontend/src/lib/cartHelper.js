export const numberOfProducts = (cart) => {
    return cart.reduce((a, b) => a + (b.product && b.quantity), 0);
}

export const calcTotalCartPrice = (cart) => {
    return cart.reduce((a, b) => a + (b.product && b.product.price * b.quantity), 0);
}

export const formatPrice = (priceInCents) => {
    const options = {
        style: 'currency',
        currency: 'USD',
        minimumFractionDigits: 2,
    };
    // if its a whole, dollar amount, leave off the .00
    if (priceInCents % 100 === 0)
        options.minimumFractionDigits = 0;

    const formatter = new Intl.NumberFormat('en-US', options);
    return formatter.format(priceInCents / 100);
}