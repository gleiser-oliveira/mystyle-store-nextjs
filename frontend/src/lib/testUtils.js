import casual from 'casual';
import { act } from 'react-dom/test-utils';
import { CURRENT_USER_QUERY } from '../components/authentication/AuthenticationOperationsGQL';

// seeded for constant results
casual.seed(123);

const fakeUser = overrides => ({
    __typename: 'User',
    id: '4234',
    name: casual.name,
    email: casual.email,
    permissions: ['USER', 'PRODUCT_CREATE'],
    cart: [],
    ...overrides,
});

const fakeUser2 = () => ({
    __typename: 'User',
    id: '2222',
    name: casual.name,
    email: casual.email,
    permissions: ['USER'],
    orders: [],
    cart: [],
});

const fakeAdminUser = () => ({
    __typename: 'User',
    id: '1111',
    name: casual.name,
    email: casual.email,
    permissions: ['ADMIN', 'PRODUCT_CREATE'],
    orders: [],
    cart: [],
});

const fakeProduct = () => ({
    __typename: 'Product',
    id: 'QWERTY0123',
    title: 'A Fake Product',
    price: 5000,
    description: 'This is a fake product for testing.',
    image: 'product.jpg',
    largeImage: 'largeProduct.jpg',
    user: fakeUser(),
});

const fakeProduct2 = () => ({
    __typename: 'Product',
    id: 'AS012',
    title: 'Another Fake Product',
    price: 2999,
    description: 'This is another fake product for testing.',
    image: 'product2.jpg',
    largeImage: 'bigProduct2.jpg',
    user: fakeUser(),
});

const fakeOrderProduct = () => ({
    __typename: 'OrderProduct',
    id: casual.uuid,
    image: `${casual.word}.jpg`,
    title: casual.words(),
    price: 4234,
    quantity: 1,
    description: casual.words(),
});

const fakeOrder = () => ({
    __typename: 'Order',
    id: 'ord123',
    charge: 'ch_123',
    total: 40000,
    products: [fakeOrderProduct(), fakeOrderProduct()],
    createdAt: 'May 1, 2020 00:00:00',
    user: fakeUser(),
});


const fakeCartProduct = overrides => ({
    __typename: 'CartProduct',
    id: 'prod123',
    quantity: 2,
    product: fakeProduct(),
    user: fakeUser(),
    ...overrides,
});

const fakeCartProduct2 = overrides => ({
    __typename: 'CartProduct',
    id: 'prod312',
    quantity: 2,
    product: fakeProduct2(),
    user: fakeUser(),
    ...overrides,
});

const cartUser = fakeUser();
const fakeCart = () => ([
    fakeCartProduct({user: cartUser}),
    fakeCartProduct2({user: cartUser}),
]);

// Fake LocalStorage
class LocalStorageMock {
    constructor() {
        this.store = {};
    }

    clear() {
        this.store = {};
    }

    getItem(key) {
        return this.store[key] || null;
    }

    setItem(key, value) {
        this.store[key] = value.toString();
    }

    removeItem(key) {
        delete this.store[key];
    }
}

// mocks
const signedInMocks = (user = fakeUser) => { return [
    {
        request: { query: CURRENT_USER_QUERY },
        result: {
            data: {
                me: user(),
            }
        }
    }
]};

const notSignedInMocks = () => ([
    {
        request: { query: CURRENT_USER_QUERY },
        result: {
            data: {
                me: null,
            }
        }
    }
]);

const updateWrapper = async (wrapper, ms = 0) => {
    await act(async () => {
        await new Promise(resolve => setTimeout(resolve, ms));
        wrapper.update();
    });
}

export {
    LocalStorageMock,
    fakeProduct,
    fakeProduct2,
    fakeUser,
    fakeUser2,
    fakeAdminUser,
    fakeCartProduct,
    fakeCartProduct2,
    fakeOrder,
    fakeCart,
    fakeOrderProduct,
    signedInMocks,
    notSignedInMocks,
    updateWrapper,
};
