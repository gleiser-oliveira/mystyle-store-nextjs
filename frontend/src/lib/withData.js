import withApollo from 'next-with-apollo';
import ApolloClient from 'apollo-boost';
import { LOCAL_STATE_QUERY } from './ClientGQL/LocalGQL';
import { DOCKER_BACKEND_URL, BROWSER_TO_BACKEND_URL } from '../config';

function createClient({ headers }) {
    // if using docker, the URI to the Yoga server must be either:
    // DOCKER_BACKEND_URL = the name of the backend server set in the docker network (used in SSR fetches)
    // BROWSER_TO_BACKEND_URL = the URL of the backend server that can be accessed from the client (used in the client's browser fetches)
    const isInsideFrontendServer = typeof window === 'undefined';
    const uri = isInsideFrontendServer ? DOCKER_BACKEND_URL : BROWSER_TO_BACKEND_URL;
    return new ApolloClient({
        uri: uri,
        request: operation => {
            operation.setContext({
                fetchOptions: {
                    credentials: 'include',
                },
                headers,
            });
        },
        // local data
        clientState: {
            resolvers: {
                Mutation: {
                    toggleCart(_, variables, client) {
                        const { cartOpen } = client.cache.readQuery({
                            query: LOCAL_STATE_QUERY,
                        });
                        // toggle the state
                        const data = {
                            data: { cartOpen: !cartOpen },
                        };
                        client.cache.writeData(data);
                        return data;
                    },
                },
                Query: {
                    localState(_, variables, client) {
                        const localState = client.cache.readQuery({
                            query: LOCAL_STATE_QUERY,
                        });
                        return localState;
                    }
                }
            },
            defaults: {
                cartOpen: false,
            }
        }
    });
}

// TODO: check getDataFromTree warning - apollo
export default withApollo(createClient);