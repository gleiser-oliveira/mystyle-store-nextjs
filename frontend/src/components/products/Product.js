import React, { Component } from 'react';
import { PropTypes } from 'prop-types';
import Link from 'next/link';
import Image from '../image/Image';
import { PriceTag, ProductStyle, Title } from './styles';
import { formatPrice } from '../../lib/cartHelper';
import DeleteProduct from './DeleteProduct';
import NeedsPermission from '../authentication/NeedsPermission';
import AddToCartButton from '../cart/AddToCartButton';

class Product extends Component {
    static propTypes = {
        product: PropTypes.object.isRequired,
    };

    render() {
        const { product } = this.props;
        const titleHref = { pathname: '/product', query: { id: product.id } };
        const editHref = { pathname: '/update', query: { id: product.id } };
        const user = this.props.userData ? this.props.userData : null;
        return (
            <ProductStyle>
                {product.image ? <Image src={product.image} alt={product.title} /> : null}
                <Title>
                    <Link href={titleHref}>
                        <a>{product.title}</a>
                    </Link>
                </Title>
                <PriceTag>{formatPrice(product.price)}</PriceTag>
                <p>{product.description}</p>
                <div className='buttonList'>
                    <NeedsPermission ownerId={product.user.id} neededPermission={['ADMIN', 'PRODUCT_UPDATE']} user={user} showMessage={false}>
                        <Link href={editHref}>
                            <a>{'\u270F'} Edit</a>
                        </Link>
                    </NeedsPermission>
                    <AddToCartButton productId={product.id} />
                    <NeedsPermission ownerId={product.user.id} neededPermission={['ADMIN', 'PRODUCT_DELETE']} user={user} showMessage={false}>
                        <DeleteProduct id={product.id}>{'\u{1F4A5}'} Delete This Product</DeleteProduct>
                    </NeedsPermission>
                </div>
            </ProductStyle>
        );
    }
}

Product.propTypes = {
    userData: PropTypes.object,
};

export default Product;
