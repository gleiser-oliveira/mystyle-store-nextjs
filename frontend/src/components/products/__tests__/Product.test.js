import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { formatPrice } from '../../../lib/cartHelper';
import Product from '../Product';
import { fakeUser, fakeProduct } from '../../../lib/testUtils';

describe('Testing <Product/> component', () => {
    it('renders the title', () => {
        const fp = fakeProduct();
        const fu = fakeUser();
        const wrapper = shallow(<Product product={fp} userData={fu}/>);
        const title = wrapper.find('styles__Title a');

        expect(title.text()).toBe(fp.title);
    });

    it('renders the image', () => {
        const fp = fakeProduct();
        const fu = fakeUser();
        const wrapper = shallow(<Product product={fp} userData={fu}/>);
        const img = wrapper.find('Image');

        expect(img.props().src).toBe(fp.image);
        expect(img.props().alt).toBe(fp.title);
    });

    it('renders the description', () => {
        const fp = fakeProduct();
        const fu = fakeUser();
        const wrapper = shallow(<Product product={fp} userData={fu}/>);
        const description = wrapper.find('p');

        expect(description.text()).toBe(fp.description);
    });

    it('renders and displays the formatted price', () => {
        const fp = fakeProduct();
        const fu = fakeUser();
        const wrapper = shallow(<Product product={fp} userData={fu}/>);

        const price = wrapper.find('styles__PriceTag');

        expect(price.text()).toBe(formatPrice(fp.price));
    });

    it('renders and matches snapshot', () => {
        const fp = fakeProduct();
        const fu = fakeUser();
        const wrapper = shallow(<Product product={fp} userData={fu}/>);

        expect(toJSON(wrapper)).toMatchSnapshot();
    });
});
