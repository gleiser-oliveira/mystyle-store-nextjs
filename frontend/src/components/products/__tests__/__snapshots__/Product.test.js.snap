// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <Product/> component renders and matches snapshot 1`] = `
<styles__ProductStyle>
  <Image
    alt="A Fake Product"
    src="product.jpg"
  />
  <styles__Title>
    <Link
      href={
        Object {
          "pathname": "/product",
          "query": Object {
            "id": "QWERTY0123",
          },
        }
      }
    >
      <a>
        A Fake Product
      </a>
    </Link>
  </styles__Title>
  <styles__PriceTag>
    $50
  </styles__PriceTag>
  <p>
    This is a fake product for testing.
  </p>
  <div
    className="buttonList"
  >
    <NeedsPermission
      neededPermission={
        Array [
          "ADMIN",
          "PRODUCT_UPDATE",
        ]
      }
      ownerId="4234"
      showMessage={false}
      user={
        Object {
          "__typename": "User",
          "cart": Array [],
          "email": "Santina_Terry@yahoo.com",
          "id": "4234",
          "name": "Mr. Evan Jenkins",
          "permissions": Array [
            "USER",
            "PRODUCT_CREATE",
          ],
        }
      }
    >
      <Link
        href={
          Object {
            "pathname": "/update",
            "query": Object {
              "id": "QWERTY0123",
            },
          }
        }
      >
        <a>
          ‚úè
           Edit
        </a>
      </Link>
    </NeedsPermission>
    <AddToCartButton
      productId="QWERTY0123"
    />
    <NeedsPermission
      neededPermission={
        Array [
          "ADMIN",
          "PRODUCT_DELETE",
        ]
      }
      ownerId="4234"
      showMessage={false}
      user={
        Object {
          "__typename": "User",
          "cart": Array [],
          "email": "Santina_Terry@yahoo.com",
          "id": "4234",
          "name": "Mr. Evan Jenkins",
          "permissions": Array [
            "USER",
            "PRODUCT_CREATE",
          ],
        }
      }
    >
      <DeleteProduct
        id="QWERTY0123"
      >
        üí•
         Delete This Product
      </DeleteProduct>
    </NeedsPermission>
  </div>
</styles__ProductStyle>
`;
